src/energy.h

```diff
@@ -242,575 +242,867 @@ class Bonded : public Energybase {
         double u = 0;
-        auto it = spc.findGroupContaining(i); // iterator to group
-        if (it != spc.groups.end()) {         // check if i belongs to group in space
-            for (size_t ig = 0; ig < spc.groups.size(); ig++) {
-                auto &g = spc.groups[ig];
-                if (&g != &(*it))         // avoid self-interaction
-                    if (not cut(g, *it))  // check g2g cut-off
-                        for (auto &j : g) // loop over particles in other group
-                            u += i2i(i, j);
-            }
-            std::ptrdiff_t i_ndx = &i - &(*(it->begin()));   // fixme c++ style
-            u += g_internal(*it, {static_cast<int>(i_ndx)}); // only int indices are used internally
-        } else {                          // particle does not belong to any group
-            for (auto &g : spc.groups) {  // i with all other *active* particles
-                for (auto &j : g) {       // (this will include only active particles)
-                    u += i2i(i, j);
+        auto &moldata = group.traits();
+        if (!moldata.rigid) {
+            if (group.atomic) {
+                // speed optimization: non-bonded interaction exclusions do not need to be checked for atomic groups
+                for (size_t i = 0; i < index; ++i) {
+                    u += particle2particle(group[index], group[i]);
+                }
+                for (size_t i = index + 1; i < group.size(); ++i) {
+                    u += particle2particle(group[index], group[i]);
+                }
+            } else {
+                for (size_t i = 0; i < index; ++i) {
+                    if (!moldata.isPairExcluded(index, i)) {
+                        u += particle2particle(group[index], group[i]);
+                    }
+                }
+                for (size_t i = index + 1; i < group.size(); ++i) {
+                    if (!moldata.isPairExcluded(index, i)) {
+                        u += particle2particle(group[index], group[i]);
+                    }
                 }
             }
         }
         return u;
     }
 
-    double i2all_parallel(const Particle &p_i) {
-        using ranges::cpp20::views::filter;
-        using ranges::cpp20::views::transform;
-
-        i_interact_with_these.clear(); // vector of particle pointers to interact with
+    /**
+     * @brief Partial internal energy of the group limited to the particles present in the index.
+     *
+     * Only such non-bonded pair interactions within the group are considered if at least one particle is present
+     * in the index. The pair exclusions defined in the molecule topology are honoured.
+     *
+     * @param group
+     * @param index internal indices of particles within the group
+     * @return energy sum between particle pairs
+     */
+    template <typename TGroup, typename TIndex> double groupInternal(const TGroup &group, const TIndex &index) {
```

Comment:
If it was a one method, then a condition at the very beginning of the method would be to discriminate if only a single particle moved to avoid that index complement computation to speed things up.